File: src/status.zig
const green = prettyzig.RGB.init(139, 233, 255); // #8ef6ff
const yellow = prettyzig.RGB.init(241, 250, 140); // #f1fa8c
const orange = prettyzig.RGB.init(255, 184, 108); // #ffb86c
const red = prettyzig.RGB.init(255, 85, 85); // #ff5555

fn contains(haystack: []const u8, needle: []const u8) bool {
    return std.mem.indexOf(u8, haystack, needle) != null;
}

pub fn display(git_data: std.AutoHashMap([]const u8, GitData)) !void {
    const stdout = std.io.getStdOut().writer();
    var git_data_iter = git_data.iterator();

    while (git_data_iter.next()) |data| {
        const status = data.value_ptr.status;
        var path_buffer: [256]u8 = undefined;
        const path = try std.fmt.bufPrint(&path_buffer, "{s} .. ", .{data.key_ptr.*});
        try prettyzig.print(stdout, path, .{});

        if (contains(status, "nothing to commit")) {
            try prettyzig.print(stdout, "CLEAN\n", .{ .color = .{ .rgb = green }, .styles = &.{.italic} });
        } else if (contains(status, "no changes added to commit")) {
            try prettyzig.print(stdout, "DIRTY (changes not added)\n", .{ .color = .{ .rgb = orange } });
        } else if (contains(status, "Changes to be committed")) {
            try prettyzig.print(stdout, "DIRTY (changes added, not committed)\n", .{ .color = .{ .rgb = red } });
        } else if (contains(status, "Your branch is ahead of")) {
            try prettyzig.print(stdout, "DIRTY (changes committed, not pushed)\n", .{ .color = .{ .rgb = red }, .styles = &.{.bold} });
        } else {
            try prettyzig.print(stdout, "UNKNOWN\n", .{ .color = .{ .rgb = yellow } });
        }
    }
}

const std = @import("std");
const prettyzig = @import("prettyzig");

pub const GitData = @import("main.zig").GitData;

File: src/status.rs
use color_eyre::{eyre::Result, owo_colors::OwoColorize};

use crate::git_data::GitData;

pub fn display(
    base_path: String,
    disable_ascii_art: bool,
    git_data: Vec<(String, GitData)>,
) -> Result<()> {
    if !disable_ascii_art {
        let ascii_art = r#"
 ________ ___  ___  ________  ________  ___  _________   
|\  _____\\  \|\  \|\_____  \|\_____  \|\  \|\___   ___\ 
\ \  \__/\ \  \\\  \\|___/  /|\|___/  /\ \  \|___ \  \_| 
 \ \   __\\ \  \\\  \   /  / /    /  / /\ \  \   \ \  \  
  \ \  \_| \ \  \\\  \ /  /_/__  /  /_/__\ \  \   \ \  \ 
   \ \__\   \ \_______\\________\\________\ \__\   \ \__\
    \|__|    \|_______|\|_______|\|_______|\|__|    \|__|
"#;

        println!("{ascii_art}");
    }

    if git_data.is_empty() {
        println!("Could not find any git repos from provided path");
        println!(
            "Make sure to add FUZZIT_BASE_PATH to your environment (ex: ~/.zshrc) or use FUZZIT_PATH before command"
        );

        return Ok(());
    }

    println!("Iterating git repos from {base_path}\n");
    for (repo_path, git_data) in git_data {
        if git_data.status.contains("nothing to commit")
            && !git_data.status.contains("branch is ahead")
        {
            if git_data.diff.is_empty() {
                println!("{repo_path} .. {}", "CLEAN".green().italic());
            } else {
                // edge case: no upstream branch, but there are local commits not pushed
                println!(
                    "{repo_path} .. {}",
                    "DIRTY (changes committed, not pushed)".red().bold()
                );
            }
        } else if git_data.status.contains("no changes added to commit")
            || git_data.status.contains("untracked")
        {
            println!(
                "{repo_path} .. {}",
                "DIRTY (changes not added)".fg_rgb::<255, 184, 108>() // orange
            );
        } else if git_data.status.contains("Changes to be committed") {
            println!(
                "{repo_path} .. {}",
                "DIRTY (changes added, not committed)".red()
            );
        } else if git_data.status.contains("branch is ahead")
            || git_data.status.contains("diverged")
        {
            println!(
                "{repo_path} .. {}",
                "DIRTY (changes committed, not pushed)".red().bold()
            );
        } else {
            println!("{repo_path} .. {}", "UNKNOWN".yellow());
        }
    }

    Ok(())
}
